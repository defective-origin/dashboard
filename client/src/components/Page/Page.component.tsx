import React from 'react'

// ---| components |---
import Layout, { LayoutProps } from 'components/Layout'
import Block from 'components/Block'
import Helmet from 'components/lib/Helmet'
import Text from 'components/lib/Text'

// ---| common |---
import { cn } from 'common/tools'

// ---| self |---
import './Page.module.scss'

export type PageProps = LayoutProps & {
  className?: string
  children?: React.ReactNode
  name?: string
  title?: React.ReactNode
  info?: React.ReactNode
  extra?: React.ReactNode
}

const NAME = 'page'

/**
 * Component description.
 *
 * How to use
 * @example
 * <Page />
 */
export function Page(props: PageProps): JSX.Element {
  const { name, title, info, extra, children, className, ...otherProps } = props
  const _className = cn(NAME, className)
  const hasHeader = title || extra || info

  return (
    <Layout className={_className} {...otherProps}>
      {/* Add page metadata */}
      <Helmet>
        { name && <title>{name}</title> }
      </Helmet>

      {hasHeader && (
        <Layout.Header className={`${NAME}__header`}>
          <Block>
            <Block.Start className={`${NAME}__header-title`}>
              <Text.H4>{title}</Text.H4>
            </Block.Start>

            <Block.Center className={`${NAME}__header-info`}>
              {info}
            </Block.Center>

            <Block.End className={`${NAME}__header-extra`}>
              {extra}
            </Block.End>
          </Block>
        </Layout.Header>
      )}

      {children}
    </Layout>
  )
}

export default Page

Page.LeftAside = Layout.LeftAside
Page.RightAside = Layout.RightAside
Page.Footer = Layout.Footer
Page.Header = Layout.Header
Page.Content = Layout.Content
Page.Block = Block
