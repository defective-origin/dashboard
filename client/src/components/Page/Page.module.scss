@import "src/Launcher/UILauncher/UILauncher.module.scss";

:global {
  .dark {
    // --text-color: white;
  }

  .light {
    // --text-color: black;
  }

  .page {
    @include stretch;

    &__header {
      padding: 1rem;

      &-title {
    
      }

      &-info {
    
      }

      &-extra {
    
      }
    }
  }
}
