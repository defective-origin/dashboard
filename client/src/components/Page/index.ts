/* INJECT_IMPORT_PLACE */

/* INJECT_DEFAULT_EXPORT_PLACE */
export { default } from './Page.component'

/* INJECT_EXPORT_PLACE */
export * from './Page.component'
