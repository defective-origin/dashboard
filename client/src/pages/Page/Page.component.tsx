import React from 'react'

// ---| core |---
import { TranslateKeys, t } from 'locale'
import { cn, react } from 'tools'

// ---| screens |---
// ---| components |---
import Text from 'components/Text'
import Portal from 'components/Portal'
import Meta, { MetaItem } from 'components/Meta'
import Actions, { ActionItem } from 'components/Actions'
import Layout, { LayoutProps } from 'components/Layout'

import Item from 'components/Item'
import Aside from 'components/Aside'
import Block from 'components/Block'
import Content from 'components/Content'
import Footer from 'components/Footer'
import Header from 'components/Header'
import Section from 'components/Section'

// ---| self |---
import css from './Page.module.scss'

export type PageMenuItem = ActionItem

export type PageProps = LayoutProps & {
  name?: TranslateKeys
  meta?: MetaItem[]
  menu?: ActionItem[]
  actions?: ActionItem[]
}

/**
 * Component description.
 *
 * How to use
 * @example
 * <Page />
 */
export function Page(props: PageProps): JSX.Element {
  const { menu = [], actions = [], name, meta, children, className, ...otherProps } = props
  const _className = cn(css.Page, className)
  const pageName = t(name)
  const tabName = t('SYSTEM.TAB_NAME', { title: pageName })

  const actionItems: ActionItem[] = actions.map(item =>
    typeof item === 'object' ? ({ ...item, v: 'text', color: 'primary' }) : item,
  ) as ActionItem[]

  return (
    <Layout className={_className} v='columns' {...otherProps}>
      <Meta title={tabName} items={meta} />

      <Portal name='page-name' content={<Text.H1 size='md' color='primary' content={pageName} />} />
      <Portal name='page-menu' content={<Actions items={menu} g='xs' size='xs' />} />
      <Portal name='page-actions' content={<Actions items={actionItems} v='y' menu='left' size='lg' />} />

      {children}
    </Layout>
  )
}

// TODO: leave only private case components? like sections? sliders? Or remove at all?
export default react.attachComponents(Page, {
  Item,
  Footer,
  Header,
  Content,
  Section,
  Block,
  Layout,
  LeftAside: Aside.Left,
  RightAside: Aside.Right,
})
