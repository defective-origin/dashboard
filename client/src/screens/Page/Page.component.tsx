import React, { useLayoutEffect } from 'react'

// ---| core |---
import { useLauncher } from 'Launcher'
import { TranslateKeys } from 'locale'

// ---| components |---
import Layout, { LayoutProps } from 'components/Layout'
import Block from 'components/Block'
import Section from 'components/Section'
import Head, { HeadItem } from 'components/Head'

// ---| common |---
import { cn } from 'common/tools'

// ---| self |---
import css from './Page.module.scss'

export type PageProps = LayoutProps & {
  className?: string
  children?: React.ReactNode
  name?: TranslateKeys | string
  meta?: HeadItem[]
}
// FIXME: add scroll to page content by default
/**
 * Component description.
 *
 * How to use
 * @example
 * <Page />
 */
export function Page(props: PageProps): JSX.Element {
  const { name, meta, children, className, ...otherProps } = props
  const _className = cn(css.Page, className)
  const app = useLauncher()
  const pageName = app.t(name as TranslateKeys)
  const tabName = app.t('SYSTEM.TAB_NAME', { title: pageName })

  useLayoutEffect(() => { app.setPageName(pageName) }, [pageName])

  return (
    <Layout className={_className} stretch {...otherProps}>
      <Head title={tabName} items={meta} />

      {children}
    </Layout>
  )
}

export default Page

Page.LeftAside = Layout.Left
Page.RightAside = Layout.Right
Page.Footer = Layout.Bottom
Page.Header = Layout.Top
Page.Content = Layout.Content
Page.Section = Section
Page.Block = Block
