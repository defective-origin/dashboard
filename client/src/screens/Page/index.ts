export { default } from './Page.component'
export * from './Page.component'
