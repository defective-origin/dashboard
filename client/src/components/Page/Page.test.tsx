// ---| tests |---
import { render } from 'tests'

// ---| self |---
import Page from './Page.component'
// import usePage from './Page.hook'
// import PageProvider, Page from './Page.context'

describe('[Page] component', () => {
  it('should render component', () => {
    const container = render(<Page />)

    expect(container.snapshot()).toMatchSnapshot()
  })
})

// describe('[Page] hook', () => {
//   it('should return options', () => {
//     const { result } = renderHook(() => usePage())
//
//     expect(result.current).toEqual(null)
//   })
// })

// describe('[Page] context', () => {
//   const wrapper = (props: PageProps) => <PageProvider { ...props } />
//
//   it('should return options', () => {
//     const { result } = renderHook(() => usePage(), { wrapper })
//
//     expect(result.current).toBeTruthy()
//   })
// })
