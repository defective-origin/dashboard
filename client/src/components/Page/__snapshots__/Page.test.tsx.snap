// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`[Page] component > should render component 1`] = `
<DocumentFragment>
  <div
    class="layout layout--columns d--y stretch _Page_d933f5"
  />
</DocumentFragment>
`;
